File: src/maze/generator.py
"""Maze generators using iterative DFS (recursive backtracker)."""
import random
from src.config import GRID_COLS, GRID_ROWS, HEX_NBR_EVEN, HEX_NBR_ODD


# ── Square maze ───────────────────────────────────────────────────────────────

def generate_square_maze(
    cols: int = GRID_COLS,
    rows: int = GRID_ROWS,
    seed: int | None = None,
) -> dict:
    """Return passages dict: {(c,r): set of connected (nc,nr)}."""
    if seed is not None:
        random.seed(seed)

    passages: dict = {(c, r): set() for r in range(rows) for c in range(cols)}
    visited  = set()
    stack    = [(0, rows - 1)]      # start from bottom-left
    visited.add((0, rows - 1))

    while stack:
        c, r = stack[-1]
        nbrs = _sq_unvisited_nbrs(c, r, cols, rows, visited)
        if nbrs:
            nc, nr = random.choice(nbrs)
            passages[(c, r)].add((nc, nr))
            passages[(nc, nr)].add((c, r))
            visited.add((nc, nr))
            stack.append((nc, nr))
        else:
            stack.pop()

    return passages


def _sq_unvisited_nbrs(c, r, cols, rows, visited):
    candidates = []
    if r > 0:        candidates.append((c, r - 1))
    if r < rows - 1: candidates.append((c, r + 1))
    if c > 0:        candidates.append((c - 1, r))
    if c < cols - 1: candidates.append((c + 1, r))
    return [n for n in candidates if n not in visited]


# ── Hex maze ──────────────────────────────────────────────────────────────────

def generate_hex_maze(
    cols: int = GRID_COLS,
    rows: int = GRID_ROWS,
    seed: int | None = None,
) -> dict:
    """Return passages dict: {(c,r): set of connected (nc,nr)}, odd-r offset."""
    if seed is not None:
        random.seed(seed)

    passages: dict = {(c, r): set() for r in range(rows) for c in range(cols)}
    visited  = set()
    stack    = [(0, rows - 1)]
    visited.add((0, rows - 1))

    while stack:
        c, r = stack[-1]
        nbrs = _hex_unvisited_nbrs(c, r, cols, rows, visited)
        if nbrs:
            nc, nr = random.choice(nbrs)
            passages[(c, r)].add((nc, nr))
            passages[(nc, nr)].add((c, r))
            visited.add((nc, nr))
            stack.append((nc, nr))
        else:
            stack.pop()

    return passages


def _hex_unvisited_nbrs(c, r, cols, rows, visited):
    offsets = HEX_NBR_ODD if r % 2 == 1 else HEX_NBR_EVEN
    candidates = []
    for dc, dr in offsets:
        nc, nr = c + dc, r + dr
        if 0 <= nc < cols and 0 <= nr < rows and (nc, nr) not in visited:
            candidates.append((nc, nr))
    return candidates

File: web/src/maze/generator.js
// Mirrors src/maze/generator.py — iterative DFS (recursive backtracker).
// Returns passages as Map<string, Set<string>> using posKey encoding.

import { GRID_COLS, GRID_ROWS, HEX_NBR_EVEN, HEX_NBR_ODD } from '../config.js';
import { posKey, seededRng, choice } from '../utils.js';

// ── Square maze ───────────────────────────────────────────────────────────────

/**
 * Generate a perfect square maze via iterative DFS.
 * @param {number} cols
 * @param {number} rows
 * @param {number|null} seed  - pass a number for reproducibility
 * @returns {Map<string, Set<string>>} passages
 */
export function generateSquareMaze(cols = GRID_COLS, rows = GRID_ROWS, seed = null) {
    const rng = (seed !== null) ? seededRng(seed) : Math.random.bind(Math);

    // Initialise every cell with an empty neighbour set
    const passages = new Map();
    for (let r = 0; r < rows; r++)
        for (let c = 0; c < cols; c++)
            passages.set(posKey(c, r), new Set());

    const visited = new Set();
    const startKey = posKey(0, rows - 1);   // bottom-left, matches Python
    visited.add(startKey);
    const stack = [startKey];

    while (stack.length > 0) {
        const key = stack[stack.length - 1];
        const [c, r] = key.split(',').map(Number);
        const nbrs = _sqUnvisited(c, r, cols, rows, visited);

        if (nbrs.length > 0) {
            const nbrKey = choice(nbrs, rng);
            // Carve passage in both directions
            passages.get(key).add(nbrKey);
            passages.get(nbrKey).add(key);
            visited.add(nbrKey);
            stack.push(nbrKey);
        } else {
            stack.pop();
        }
    }

    return passages;
}

function _sqUnvisited(c, r, cols, rows, visited) {
    const candidates = [];
    if (r > 0)        candidates.push(posKey(c,     r - 1));
    if (r < rows - 1) candidates.push(posKey(c,     r + 1));
    if (c > 0)        candidates.push(posKey(c - 1, r));
    if (c < cols - 1) candidates.push(posKey(c + 1, r));
    return candidates.filter(k => !visited.has(k));
}

// ── Hex maze ──────────────────────────────────────────────────────────────────

/**
 * Generate a perfect hex maze via iterative DFS (odd-r offset coords).
 * @param {number} cols
 * @param {number} rows
 * @param {number|null} seed
 * @returns {Map<string, Set<string>>} passages
 */
export function generateHexMaze(cols = GRID_COLS, rows = GRID_ROWS, seed = null) {
    const rng = (seed !== null) ? seededRng(seed) : Math.random.bind(Math);

    const passages = new Map();
    for (let r = 0; r < rows; r++)
        for (let c = 0; c < cols; c++)
            passages.set(posKey(c, r), new Set());

    const visited  = new Set();
    const startKey = posKey(0, rows - 1);
    visited.add(startKey);
    const stack = [startKey];

    while (stack.length > 0) {
        const key = stack[stack.length - 1];
        const [c, r] = key.split(',').map(Number);
        const nbrs = _hexUnvisited(c, r, cols, rows, visited);

        if (nbrs.length > 0) {
            const nbrKey = choice(nbrs, rng);
            passages.get(key).add(nbrKey);
            passages.get(nbrKey).add(key);
            visited.add(nbrKey);
            stack.push(nbrKey);
        } else {
            stack.pop();
        }
    }

    return passages;
}

function _hexUnvisited(c, r, cols, rows, visited) {
    const offsets = (r % 2 === 1) ? HEX_NBR_ODD : HEX_NBR_EVEN;
    const candidates = [];
    for (const [dc, dr] of offsets) {
        const nc = c + dc, nr = r + dr;
        if (nc >= 0 && nc < cols && nr >= 0 && nr < rows) {
            const k = posKey(nc, nr);
            if (!visited.has(k)) candidates.push(k);
        }
    }
    return candidates;
}
